pkg/apis/v1alpha1/stage_types.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	StageKind = "Stage"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Stage is an API that describes the staged change of a resource
+type Stage struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// Spec holds information about the request being evaluated.
+	Spec StageSpec `json:"spec,omitempty"`
+}
+
+// StageSpec defines the specification for Stage.
+type StageSpec struct {
+	// ResourceRef specifies the Kind and version of the resource.
+	ResourceRef StageResourceRef `json:"resourceRef"`
+	// Selector specifies the stags will be applied to the selected resource.
+	Selector *StageSelector `json:"selector,omitempty"`
+	// Weight means the current stage, in case of multiple stages,
+	// a random stage will be matched as the next stage based on the weight.
+	// +default=0
+	Weight int `json:"weight,omitempty"`
+	// Delay means there is a delay in this stage.
+	Delay *StageDelay `json:"delay,omitempty"`
+	// Next indicates that this stage will be moved to.
+	Next StageNext `json:"next"`
+}
+
+// StageResourceRef specifies the kind and version of the resource.
+type StageResourceRef struct {
+	// APIGroup of the referent.
+	// +default="v1"
+	APIGroup string `json:"apiGroup,omitempty"`
+	// Kind of the referent.
+	Kind string `json:"kind"`
+}
+
+// StageDelay describes the delay time before going to next.
+type StageDelay struct {
+	// DurationSeconds indicates the stage delay time (in seconds).
+	// If JitterDurationSeconds is less than DurationSeconds, then JitterDurationSeconds is used.
+	DurationSeconds int `json:"durationSeconds"`
```

Comment:
IMO Seconds -> MilliSeconds would be more fine-grained.  (applies to other XYZSeconds as well)

(the upstream also met this issue due to hardcode to seconds in first place: https://github.com/kubernetes/kubernetes/issues/76951)

Reply:
I don't have a strong opinion on this issue.
If we use `MilliSeconds`, should we add both `Seconds` and `MilliSeconds` as they do upstream?

In my opinion, maybe it's more generic if we just use `time.Duration`?

``` go
Duration time.Duration `json:"duration"`
```

``` yaml
delay:
  duration: 1s
```

Reply:
> should we add both Seconds and MilliSeconds as they do upstream?

Nope, just MilliSeconds. The upstream adds both for backward compatibility.

Reply:
> In my opinion, maybe it's more generic if we just use time.Duration?

Practically, it's _ok_ if we stick it to a config option. However, if we want better extensibility (like implement in CRD) and non-Go support, we'd better follow the abovementioned API conventions, in particular:

> Duration fields must be represented as integer fields with units being part of the field name (e.g. leaseDurationSeconds). We don't use Duration in the API since that would require clients to implement go-compatible parsing.

Reply:
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#units

I get it, it sounds very reasonable, I will add it in the future.

Reply:
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#units